main/src/ca/uwaterloo/flix/language/phase/EntryPoint.scala

```diff
@@ -0,0 +1,295 @@
+
+    mapN(argsActionVal, resultActionVal) {
+      case (argsAction, resultAction) =>
+        mkEntryPoint(defn, argsAction, resultAction, root)
+    }
+  }
+
+  /**
+    * Checks the entry point function arguments.
+    * Returns a flag indicating whether the args should be passed to this function or ignored.
+    */
+  private def checkEntryPointArgs(defn: TypedAst.Def, classEnv: Map[Symbol.ClassSym, Ast.ClassContext])(implicit flix: Flix): Validation[EntryPointArgsAction, EntryPointError] = defn match {
+    case TypedAst.Def(sym, TypedAst.Spec(_, _, _, _, _, declaredScheme, _, _, _), _) =>
+      val unitSc = Scheme.generalize(Nil, Type.Unit)
+
+      // First check that there's exactly one argument.
+      val argVal = declaredScheme.base.arrowArgTypes match {
+        // Case 1: One arg. Ok :)
+        case arg :: Nil => arg.toSuccess
+        // Case 2: Multiple args. Error.
+        case _ :: _ :: _ => EntryPointError.TooManyEntryPointArgs(sym, sym.loc).toFailure
+        // Case 3: Empty arguments. Impossible since this is desugared to Unit.
+        case Nil => throw InternalCompilerException("Unexpected empty argument list.")
+      }
+
+      flatMapN(argVal: Validation[Type, EntryPointError]) {
+        arg =>
+          val argSc = Scheme.generalize(Nil, arg)
+          val stringArraySc = Scheme.generalize(Nil, Type.mkArray(Type.Str, SourceLocation.Unknown))
+
+          if (Scheme.equal(unitSc, argSc, classEnv)) {
+            // Case 1: Unit -> XYZ. We can ignore the args.
+            EntryPointArgsAction.Ignore.toSuccess
+          } else if (Scheme.equal(argSc, stringArraySc, classEnv)) {
+            // Case 2: Array[String] -> XYZ. We need to pass along the input args.
+            EntryPointArgsAction.Use.toSuccess
+          } else {
+            // Case 3: Bad arguments. Error.
+            EntryPointError.UnexpectedEntryPointArg(sym, arg, sym.loc).toFailure
+          }
+      }
+  }
+
+  /**
+    * Checks the entry point function result type.
+    * Returns a flag indicating whether the result should be printed, cast, or unchanged.
+    */
+  private def checkEntryPointResult(defn: TypedAst.Def, root: TypedAst.Root, classEnv: Map[Symbol.ClassSym, Ast.ClassContext])(implicit flix: Flix): Validation[EntryPointResultAction, EntryPointError] = defn match {
+    case TypedAst.Def(sym, TypedAst.Spec(_, _, _, _, _, declaredScheme, _, _, _), _) =>
+      val resultTpe = declaredScheme.base.arrowResultType
+      val unitSc = Scheme.generalize(Nil, Type.Unit)
+      val resultSc = Scheme.generalize(Nil, resultTpe)
+
+      val toString = root.classes(new Symbol.ClassSym(Nil, "ToString", SourceLocation.Unknown)).sym
+
+      if (Scheme.equal(unitSc, resultSc, classEnv)) {
+        // Case 1: XYZ -> Unit.
+        if (declaredScheme.base.arrowEffectType == Type.Pure) {
+          // Case 1.1: XYZ ->{Pure} Unit. Need to cast.
+          EntryPointResultAction.Cast.toSuccess
```

Comment:
Why do we need to cast? Calling print is impure.

Reply:
Ah, because if its unit and we don't print?

Reply:
And we can't just _always_ cast because of redundancy errors.

Reply:
:O